File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.js
function solution(n, edge) {
    let obj = {};
    let q = [[1, 0]];
    let dist = new Array(n).fill(0); dist[0] = 0;
    let visited = new Set();
    visited.add(1);
    for (let i = 0; i < edge.length; i++) {
        let [n1, n2] = edge[i];
        if (!obj[n1]) obj[n1] = [];
        if (!obj[n2]) obj[n2] = [];
        obj[n1].push(n2);
        obj[n2].push(n1);
    }
    while (q.length) {
        let [node, d] = q.shift();
        obj[node].forEach((e => {
            if (visited.has(e)) return;
            q.push([e, d + 1])
            dist[e - 1] = d + 1;
            visited.add(e);
        }));
    }
    let max = Math.max(...dist);
    return dist.reduce((acc, cur) => {
        if (cur == max) return acc + 1;
        return acc;
    }, 0);
}

File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.py
from collections import deque
def solution(n, edge):
    dic = {}
    answer = 0
    q = deque([[1, 0]])
    visited = set([1])
    dist = [0] * n
    for e in edge:
        [n1, n2] = e
        if not n1 in dic: dic[n1] = []
        if not n2 in dic: dic[n2] = []
        dic[n1].append(n2)
        dic[n2].append(n1)
    while len(q):
        [node, d] = q.popleft()
        for n in dic[node]:
            if n in visited: continue
            q.append([n, d + 1])
            dist[n - 1] = d + 1
            visited.add(n)
    maximum = max(dist)
    for d in dist:
        if d == maximum:
            answer += 1
    return answer